### ISSUE TITLE
feat(learning_rate): Add CyclicLR — triangular cycling (Smith 2017)

### ISSUE THREAD

**Issue by sanskarpan** [sanskarpan/Optimization#34]
## Description
Cycles learning rate between `base_lr` and `max_lr` with three modes.

```
cycle = floor(1 + step / (2*step_size))
x = |step/step_size - 2*cycle + 1|
triangular:  lr = base_lr + (max_lr-base_lr)*max(0,1-x)
triangular2: lr = ... / 2^(cycle-1)           # halving amplitude
exp_range:   lr = ... * gamma^step             # exponential decay
```

## Acceptance Criteria
- [ ] All three modes implemented
- [ ] LR always in [base_lr, max_lr]
- [ ] Invalid mode raises `ValueError`; step_size<=0 raises `ValueError`
